cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -129,29 +131,25 @@ public void addNotification(ProgramSchedule schedule, Notification notification)
   private void addNotification(Job job, Notification notification) {
     List<Notification> notifications = new ArrayList<>(job.getNotifications());
     notifications.add(notification);
-    Job newJob = new SimpleJob(job.getSchedule(), job.getCreationTime(), notifications, job.getState());
-    if (isTriggerSatisfied(newJob)) {
-      newJob = doTransitState(job, Job.State.PENDING_CONSTRAINT);
+
+    Job.State newState = job.getState();
+    if (isTriggerSatisfied(job.getSchedule().getTrigger(), notifications)) {
+      newState = Job.State.PENDING_CONSTRAINT;
+      job.getState().checkTransition(newState);
```

Comment:
I think its ok to have this check as a sanity check.
It'll help enforce the state transitions in the future, in case someone tries to refactor this code.